src/Stratis.Bitcoin/Utilities/ParallelAsync.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    public static class ParallelAsync
+    {
+        /// <summary>
+        /// Executes a foreach operation on an IEnumerable in which iterations run in parallel.
+        /// </summary>
+        /// <typeparam name="TSource">Item type.</typeparam>
+        /// <param name="collection">Enumerated collection.</param>
+        /// <param name="maxDegreeOfParallelism">The maximum amount of items that can be processed simultaneously.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <param name="action">Action that is used for processing each item in the collection.</param>
+        public static async Task ForEachAsync<TSource>(this IEnumerable<TSource> collection, int maxDegreeOfParallelism, CancellationToken cancellationToken, Func<TSource, Task> action)
+        {
+            var unfinished = new List<Task>();
+
+            foreach (TSource item in collection)
+            {
+                if (cancellationToken.IsCancellationRequested)
```

Comment:
@noescape00 Correct me if I'm wrong... if we cancel here... then won't we still execut all the unfinished tasks here?

`await Task.WhenAll(unfinished).ConfigureAwait(false);`

I would assume that if the node shuts down that we stop processing tasks? In context of peer discovery, if the nodes disposes, we want to stop peer discovery immediately.

Maybe if cancellation is requested, the we just clear the unfinished list?

Reply:
>This is really cool @noescape00 and WELL SPOTTED dude! Good job

Thanks! 🙂 

>Maybe if cancellation is requested, the we just clear the unfinished list?

well, they still will be running, we just won't await them and doing that might be dangerous- we can end up disposing something and then the task executes and bumps into a disposed object. 

I think the tasks themselves should be created with the consideration of the cancellation.

Reply:
hm, maybe to make it more obvious that tasks should consider the cancellation I should make an overload that will provide the cancellation instead of making tasks using tokens from the outside.

Reply:
Done.

Reply:
I wonder if we may want to catch OperationCancellation exception on the awaits (especially the `WhenAny`) as you might end up not waiting for some tasks to complete.

Reply:
WhenAny doesn't rethrow the exception so in case one of the tasks throws- it's still perfectly fine and the rest of the tasks will be awaited.

Reply:
And as for the WhenAll- here I can use a foreach loop that will await each task in the list and if there is an exception- it will be ignored. 
Would you like me to do that?

Reply:
No that's fine as the WhenAll will wait for all tasks to complete.

Reply:
Ah, yes, you're right. Missed that 🙂